### ISSUE TITLE
WHY MAE

### ISSUE THREAD

**Issue by VLadImirluren** [facebookresearch/sam2#594]
Solid work!

I am curious about the selection of image encoder. Why did you adopt MAE as your image encoder? Why not DINO?

In my opinion, semantic segmentation heavily rely on semantic information, but MAE is trained as an autoencoder,
which intuitively lacked of semantic information. 

In contrast, DINO feature is widely used, which can be used to predict semantic map, depth map, normal map.....

So, why not DINO? Why MAE?

Thanks

**Comment by nam-drun** [comment]
DINO's backbone is based on ViT so I will begin my guesstimate from ViT. For starters, table 5 from "Three things everyone should know about Vision Transformers" indicated that only finetuning Multi-Head Self-Attention not only saves 66% ViT's storage on VRAM but ended up giving more accuracy. So despite MLP holds 2/3 parameter counts in ViT, it is Self-Attention's performance that did the most performance. I guess who would have thought, still an empirical result to ground my guesstimate is recommended. 

Let's begin my long guesstimate with Multi-Head Self-Attention. 

First consider @ as matrix multiplication and passing input to FC layer with weight W is the same as x @ W. Based on Timm's ViT implementation, Self-Attention has 4 phases:
1. Q = x @ W_Q; K = x @ W_K; V = x @ W_V 
2. LayerNorm(Q); LayerNorm(K). This is a regularising factor.
3. attn_score = Softmax(Q @ K/ sqrt(head_dim). Side note, scaling down by sqrt(head_dim) is sort of like BatchNorm but instead of scaling down entire minibatch by the batch's mean, you scale it down by the fixed number that happens to be the size of embedding dimension for each attention head. So this can be considered as a regularising factor.
4. Output = Q @ V. If you expand (4) into attn_score @ (x @ W_V), you can rewrite it as (attn_score @ x) @ W_V because linear algebra allows it but sure I will check on Google Colab note.

Observation 1: Comparing to MLP's operation which is Output = Softmax(matmul(input, W)). It's naturally to think that Self-Attention as 2 parallel MLP's intermediate outputs (Q, K) regularised by LayerNorm and BatchNorm's esque sqrt(head_dim). From this recent paper "Attention Guided CAM: Visual Explanations of Vision Transformer Guided by Self-Attention" which visualises attention_score, you can see attn_score alone already attended what is relevant in the input image. So if ***Query and Key alone can already told where to attend by how much in input image***, what use is the 4th phase? From the equivalent equation I write in 4th phase, ***essentially Value in Self-Attention is akin to passing Attention Score into FC layer and skip-connection (mutiply-wise) with the intermediate outout, or a context-aware linear classifier***. Given FC layer "summarises information" from what I recall from uni lecture unless Output Channel > Input Channel, which is never the case afaik from looking at various Vision Transformer config, this also implicates that V is acting as a context-aware downsampler from the next layer (though not the same operation as Downsampling layer in CNN). So TLDR, ***Self-Attention is 2 parallel MLP with 2 regularising factors and 1 context-aware linear classifier/downsampler***.

Guesstimate 1: ***If Self-Attention has many MLP bits, then it also shares the same weakness like MLP but a lesser extent when regarding about local inductive bias***. Refer to Table 1 in "Studying inductive biases in image classification task" paper and "Finding the Needle in the Haystack with Convolutions: on the benefits of architectural bias" pretty much confirm that but also give insight 1 important thing: Researchers have tried to solve MLP's problem (despite its expressiveness than Convolution) by forcefully limiting its receptive field, similar-ish to why Swin was built. Now here's the kicker, if you forcefully limit receptive field to attend like CNN, then you share the same weakness of global receptive in each Convolution layer. Impossible to attend sparsely separated multiple fine-details sparsely far apart in the input and having to rely on deep network to regain global receptive; and deep architecture has its own problem with convergence rate that led to skip-connection being introduced in the 1st place etc etc. Which leads to the intuition being SimMIM and MAE

==== Why MAE over DinoV2 ====
Guesstimate 2: ***Based on Guesstimate 1, by keeping 25% input patches at sparse locations, random_masking module in MAE forces Self-Attention block to attend only very, very few fine-details, but these fine-details can be anywhere in the input***. To sanity-check my guesstimate, I refer to "Improving Masked Autoencoders by Learning Where to Mask" and "SimMIM: a Simple Framework for Masked Image Modeling". 

Guesstimate 3: Scaling Vision Transformers to 22 Billion Parameters stated that: "In scaling ViT beyond prior works, we observed divergent training loss after a few thousand steps. In particular, this instability was observed for models with around 8B parameters (see Appendix B). It was caused by extremely large values in attention logits, which lead to (almost one-hot)  attention weights with near-zero entropy. To solve this, we adopt the approach of Gilmer et al. (2023), which applies LayerNorm (Ba et al., 2016) to the queries and keys before the dot-product attention computation." Not only this paper but most modern days Vision Transformer tend to this but apparently that wasn't enough because you can still see in Attention Guided CAM which was published 1 years after : Scaling Vision Transformers to 22B. ***By removing 75% of input patch, you also reduce the chance of this phenomenon happens.***

Sorry for making you reflect back from DINOv2 -> ViT -> Self-Attention -> Attention Score, then jump straight to MAE.

**Comment by VLadImirluren** [comment]
> DINO's backbone is based on ViT so I will begin my guesstimate from ViT. For starters, table 5 from "Three things everyone should know about Vision Transformers" indicated that only finetuning Multi-Head Self-Attention not only saves 66% ViT's storage on VRAM but ended up giving more accuracy. So despite MLP holds 2/3 parameter counts in ViT, it is Self-Attention's performance that did the most performance. I guess who would have thought, still an empirical result to ground my guesstimate is recommended.
> 
> Let's begin my long guesstimate with Multi-Head Self-Attention.
> 
> First consider @ as matrix multiplication and passing input to FC layer with weight W is the same as x @ W. Based on Timm's ViT implementation, Self-Attention has 4 phases:
> 
> 1. Q = x @ W_Q; K = x @ W_K; V = x @ W_V
> 2. LayerNorm(Q); LayerNorm(K). This is a regularising factor.
> 3. attn_score = Softmax(Q @ K/ sqrt(head_dim). Side note, scaling down by sqrt(head_dim) is sort of like BatchNorm but instead of scaling down entire minibatch by the batch's mean, you scale it down by the fixed number that happens to be the size of embedding dimension for each attention head. So this can be considered as a regularising factor.
> 4. Output = Q @ V. If you expand (4) into attn_score @ (x @ W_V), you can rewrite it as (attn_score @ x) @ W_V because linear algebra allows it but sure I will check on Google Colab note.
> 
> Observation 1: Comparing to MLP's operation which is Output = Softmax(matmul(input, W)). It's naturally to think that Self-Attention as 2 parallel MLP's intermediate outputs (Q, K) regularised by LayerNorm and BatchNorm's esque sqrt(head_dim). From this recent paper "Attention Guided CAM: Visual Explanations of Vision Transformer Guided by Self-Attention" which visualises attention_score, you can see attn_score alone already attended what is relevant in the input image. So if _**Query and Key alone can already told where to attend by how much in input image**_, what use is the 4th phase? From the equivalent equation I write in 4th phase, _**essentially Value in Self-Attention is akin to passing Attention Score into FC layer and skip-connection (mutiply-wise) with the intermediate outout, or a context-aware linear classifier**_. Given FC layer "summarises information" from what I recall from uni lecture unless Output Channel > Input Channel, which is never the case afaik from looking at various Vision Transformer config, this also implicates that V is acting as a context-aware downsampler from the next layer (though not the same operation as Downsampling layer in CNN). So TLDR, _**Self-Attention is 2 parallel MLP with 2 regularising factors and 1 context-aware linear classifier/downsampler**_.
> 
> Guessimate 1: _**If Self-Attention has many MLP bits, then it also shares the same weakness like MLP but a lesser extent when regarding about local inductive bias**_. Refer to Table 1 in "Studying inductive biases in image classification task" paper and "Finding the Needle in the Haystack with Convolutions: on the benefits of architectural bias" pretty much confirm that but also give insight 1 important thing: Researchers have tried to solve MLP's problem (despite its expressiveness than Convolution) by forcefully limiting its receptive field, similar-ish to why Swin was built. Now here's the kicker, if you forcefully limit receptive field to attend like CNN, then you share the same weakness of global receptive in each Convolution layer. Impossible to attend sparsely separated multiple fine-details sparsely far apart in the input and having to rely on deep network to regain global receptive; and deep architecture has its own problem with convergence rate that led to skip-connection being introduced in the 1st place etc etc. Which leads to the intuition being SimMIM and MAE
> 
> ==== Why MAE over DinoV2 ==== Guesstimate 2: _**Based on Guesstimate 1, by keeping 25% input patches at sparse locations, random_masking module in MAE forces Self-Attention block to attend only very, very few fine-details, but these fine-details can be anywhere in the input**_. To sanity-check my guesstimate, I refer to "Improving Masked Autoencoders by Learning Where to Mask" and "SimMIM: a Simple Framework for Masked Image Modeling".
> 
> Guesstimate 3: Scaling Vision Transformers to 22 Billion Parameters stated that: "In scaling ViT beyond prior works, we observed divergent training loss after a few thousand steps. In particular, this instability was observed for models with around 8B parameters (see Appendix B). It was caused by extremely large values in attention logits, which lead to (almost one-hot) attention weights with near-zero entropy. To solve this, we adopt the approach of Gilmer et al. (2023), which applies LayerNorm (Ba et al., 2016) to the queries and keys before the dot-product attention computation." Not only this paper but most modern days Vision Transformer tend to this but apparently that wasn't enough because you can still see in Attention Guided CAM which was published 1 years after : Scaling Vision Transformers to 22B. _**By removing 75% of input patch, you also reduce the chance of this phenomenon happens.**_
> 
> Sorry for making you reflect back from DINOv2 -> ViT -> Self-Attention -> Attention Score, then jump straight to MAE.

THANKS!!!!!!!!!!

**Comment by nam-drun** [comment]
> THANKS!!!!!!!!!!

No problem. My paper summary between DINOv2 vs MAE is that:
- DINOv2 changes training signal to incentivise backbone ViT learning to be local inductive bias like CNN but still retain its long-term context-aware inherent nature.
- random_masking module in MAE literally deletes majority of input image so that backbone ViT has no other choice than learning to be local inductive bias like CNN but luckily it doesn't delete too much so backbone ViT still retain its long-term context-aware inherent nature. In a metaphor, MAE is "necessity is the mother of invention", under MAE's pipeline, backbone ViT learnt to be better because it had to.
